Lib/test/test_contextlib.py

```diff
@@ -1080,5 +1081,47 @@ def test_cm_is_reentrant(self):
             1/0
         self.assertTrue(outer_continued)
 
+
+class TestChdir(unittest.TestCase):
+    def test_simple(self):
+        old_cwd = os.getcwd()
+        target = os.path.join(os.path.dirname(__file__), 'data')
+        assert old_cwd != target
+
+        with chdir(target):
+            assert os.getcwd() == target
+        assert os.getcwd() == old_cwd
+
+    def test_reentrant(self):
+        old_cwd = os.getcwd()
+        target1 = os.path.join(os.path.dirname(__file__), 'data')
+        target2 = os.path.join(os.path.dirname(__file__), 'ziptestdata')
+        assert old_cwd not in (target1, target2)
+        chdir1, chdir2 = chdir(target1), chdir(target2)
+
+        with chdir1:
+            assert os.getcwd() == target1
+            with chdir2:
+                assert os.getcwd() == target2
+                with chdir1:
+                    assert os.getcwd() == target1
+                assert os.getcwd() == target2
+            assert os.getcwd() == target1
+        assert os.getcwd() == old_cwd
+
+    def test_exception(self):
+        old_cwd = os.getcwd()
+        target = os.path.join(os.path.dirname(__file__), 'data')
+        assert old_cwd != target
+
+        try:
+            with chdir(target):
+                assert os.getcwd() == target
+                raise RuntimeError()
+        except RuntimeError:
```

Comment:
Can you create your own exception class here instead of using a builtin exception class?